networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/subnets/NodeBasedStableSubnetSubscriberTest.java

```diff
@@ -0,0 +1,153 @@
+        .thenReturn(List.of(UInt64.valueOf(12), UInt64.valueOf(63)))
+        .thenReturn(List.of(UInt64.valueOf(12), UInt64.valueOf(25)));
+
+    // Subnet 12 expires at slot 8 and subnet 63 at slot 4
+    when(miscHelpersMock.calculateNodeSubnetUnsubscriptionSlot(any(), eq(UInt64.ZERO)))
+        .thenReturn(UInt64.valueOf(8))
+        .thenReturn(UInt64.valueOf(4));
+    // Subnet 12 and 25 of the next round expire both at slot 12
+    when(miscHelpersMock.calculateNodeSubnetUnsubscriptionSlot(any(), eq(UInt64.valueOf(5))))
+        .thenReturn(UInt64.valueOf(12));
+
+    when(specVersionMock.miscHelpers()).thenReturn(miscHelpersMock);
+    when(specMock.atSlot(any())).thenReturn(specVersionMock);
+    NetworkingSpecConfig networkingSpecConfigMock = mock(NetworkingSpecConfig.class);
+    when(networkingSpecConfigMock.getSubnetsPerNode()).thenReturn(2);
+    when(specMock.getNetworkingConfig()).thenReturn(networkingSpecConfigMock);
+
+    final AttestationTopicSubscriber attestationTopicSubscriberMock =
+        mock(AttestationTopicSubscriber.class);
+
+    final NodeBasedStableSubnetSubscriber subscriber =
+        new NodeBasedStableSubnetSubscriber(
+            attestationTopicSubscriberMock,
+            specMock,
+            Optional.of(dataStructureUtil.randomUInt256()));
+
+    final ArgumentCaptor<Set<SubnetSubscription>> subscriptionCaptor =
+        ArgumentCaptor.forClass(Set.class);
+
+    // Should subscribe to subnet 12 until slot 8 and subnet 63 until slot 4
+    subscriber.onSlot(UInt64.ZERO);
+    verify(attestationTopicSubscriberMock, times(1))
+        .subscribeToPersistentSubnets(subscriptionCaptor.capture());
+    Set<SubnetSubscription> actual = subscriptionCaptor.getValue();
+    assertThat(actual).hasSize(2);
+    assertThat(actual).contains(new SubnetSubscription(12, UInt64.valueOf(8)));
+    assertThat(actual).contains(new SubnetSubscription(63, UInt64.valueOf(4)));
+
+    // Subnet 63 expires, subnet 12 is still active
+    // Subnet 12 expiration should be updated to slot 12
+    // subscribe to subnet 25 until slot 12
+    subscriber.onSlot(UInt64.valueOf(5));
+    verify(attestationTopicSubscriberMock, times(2))
+        .subscribeToPersistentSubnets(subscriptionCaptor.capture());
+    actual = subscriptionCaptor.getValue();
+    assertThat(actual).hasSize(2);
+    assertThat(actual).contains(new SubnetSubscription(12, UInt64.valueOf(12)));
+    assertThat(actual).contains(new SubnetSubscription(25, UInt64.valueOf(12)));
+  }
+
+  @NotNull
+  private NodeBasedStableSubnetSubscriber createSubscriber() {
+    return new NodeBasedStableSubnetSubscriber(
+        new AttestationTopicSubscriber(spec, network),
+        spec,
+        Optional.of(dataStructureUtil.randomUInt256()));
+  }
+
+  private void assertSubscribedToAtLeastOneNewSubnet(List<Integer> subscribedSubnetIds) {
+    IntSet subnetIds =
+        IntOpenHashSet.toSet(subscribedSubnetIds.stream().mapToInt(Integer::intValue));
```

Comment:
why so difficult? it breaks the brain a little looking for extra logic except converting List to the set

Reply:
Removed it. Doesn't look relevant to me